packages/protocol/contracts/L1/libs/LibL2Tokenomics.sol

```diff
@@ -0,0 +1,127 @@
+    error L1_1559_Y_SCALE_TOO_LARGE();
+    error L1_OUT_OF_BLOCK_SPACE();
+
+    function getL2Basefee(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        uint32 gasLimit
+    ) internal view returns (uint64 basefee, uint64 newGasExcess) {
+        if (config.gasIssuedPerSecond == 0) {
+            // L2 1559 disabled
+            return (0, 0);
+        }
+
+        unchecked {
+            uint256 reduced = (block.timestamp - state.lastProposedAt) *
+                config.gasIssuedPerSecond;
+            newGasExcess = uint64(reduced.max(state.l2GasExcess) - reduced);
+        }
+
+        basefee = calcL2Basefee({
+            l2GasExcess: newGasExcess,
+            xscale: state.l2Xscale,
+            yscale: uint256(state.l2Yscale) << 64,
+            gasAmount: gasLimit
+        }).toUint64();
+
+        newGasExcess += gasLimit;
+    }
+
+    function calcL2BasefeeParams(
+        uint64 gasExcessMax,
+        uint64 basefeeInitial,
+        uint64 gasTarget,
+        uint64 expected2X1XRatio
+    ) internal pure returns (uint64 l2GasExcess, uint64 xscale, uint64 yscale) {
+        assert(gasExcessMax != 0);
+
+        l2GasExcess = gasExcessMax / 2;
+
+        // calculate xscale
+        uint256 _xscale = LibFixedPointMath.MAX_EXP_INPUT / gasExcessMax;
+        if (_xscale >= type(uint64).max) {
+            revert L1_1559_X_SCALE_TOO_LARGE();
+        }
+        xscale = uint64(_xscale);
+
+        // calculate yscale
+        uint256 _yscale = calcL2Basefee(
+            l2GasExcess,
+            xscale,
+            basefeeInitial,
+            gasTarget
+        );
+        if ((_yscale >> 64) >= type(uint64).max) {
+            revert L1_1559_Y_SCALE_TOO_LARGE();
+        }
+
+        yscale = uint64(_yscale >> 64);
+
+        // Verify the gas price ratio between two blocks, one has
+        // 2*gasTarget gas and the other one has gasTarget gas.
+        {
+            _yscale = uint256(yscale) << 64;
+            uint256 price1x = calcL2Basefee(
+                l2GasExcess,
+                xscale,
+                _yscale,
+                gasTarget
+            );
+            uint256 price2x = calcL2Basefee(
+                l2GasExcess,
+                xscale,
+                _yscale,
+                gasTarget * 2
+            );
+
+            uint64 ratio = uint64((price2x * 100) / price1x);
+
+            if (expected2X1XRatio != ratio) {
+                revert L1_1559_GAS_CHANGE_MISMATCH(expected2X1XRatio, ratio);
+            }
+        }
+    }
+
+    function calcL2Basefee(
+        uint64 l2GasExcess,
+        uint64 xscale,
+        uint256 yscale,
+        uint64 gasAmount
+    ) internal pure returns (uint256) {
+        assert(gasAmount != 0 && xscale != 0 && yscale != 0);
+        uint256 _before = _ethqty(l2GasExcess, xscale);
+        uint256 _after = _ethqty(l2GasExcess + gasAmount, xscale);
+        return (_after - _before) / gasAmount / yscale;
```

Comment:
This is the spot base fee, when more gas are purchased, the basefee will go up and depends on the amount of gas purchased. I asked him this question.

Reply:
Ah I see, you're calculating the "implied average basefee", not the final basefee! Because this currently uses `gasLimit` I think this works, but I don't think this will still work when we switch to `gas_used` (see also the things Vitalik wrote at the end of his post).

The problem is because of a circular dependency: we don't know what `gas_used` is before executing all transactions, and we can only know what transactions will be valid and how they will execute when we know `base_fee` (the tx owner needs to be able to pay the ETH fee, tx.max_basefee >= block.basefee, basefee also available in smart contracts)! 

So following what Vitalik wrote (the second option), we need to set the basefee to something deterministic which for us could be

```
basefee = eth_qty(excess_gas_issued_before_block + gasLimit) / (TARGET * ADJUSTMENT_QUOTIENT)
```

We execute the block with that `basefee` and then when we know `gas_used`. We can now actually update `excess_gas_issued` correctly and refund tx owners with the "implied average basefee".

His first option seems simpler (just send everything to the proposer regardless of basefee and let the proposer pay), but I'm not sure what basefee should be used in this case (the max one in the tx would be bad) and so seems similar to the 2nd option to me. Also depending how it works exactly, it may be possible that the amount of fees collected < amount to burn, which would make the complete block invalid (which I guess would be fine).